### PR TITLE
Shivers: Add events and fix require puzzle hints logic

### PR DESCRIPTION
**Pull request by korydondzila**
## What is this fixing or adding?
Adds events to help clean up some logic and fixes require puzzle hints so that entrances are also accounted for and adds missing logic.
Also just some general cleanup.

Added a unique item that updates every year, id remains unique for each rename. This is a goal item that I wanted to show up for a small bit of in game humor.

## How was this tested?
Ran multiple generations to make sure nothing failed and played a few games solo and with others, including GodlFire the existing world maintainer, to make sure the logic was correct.

## New Maintainer
I have made a few small prior changes to Shivers AP code.
https://github.com/ArchipelagoMW/Archipelago/pull/2690
https://github.com/ArchipelagoMW/Archipelago/pull/2869
https://github.com/ArchipelagoMW/Archipelago/pull/3558
https://github.com/ArchipelagoMW/Archipelago/pull/3854

Though I have primarily been a maintainer for the Shivers client.
https://github.com/GodlFire/Shivers-Randomizer-CSharp/pulls?q=is%3Apr+author%3Akorydondzila+

I was asked by GodlFire to figure out how to get events working so that redundancy could be removed and logic improved. Also to remove the pot duplicates from the item pool so they don't show up in the advanced options. (these items are solely used to indicate where each pot goes when received but are important for progression)

### CODE REVIEW

**Comment by Exempt-Medic** [comment]
To quote the docs regarding adding a world maintainer:
> To help the core team review the change, information about the new maintainer and their contributions should be included in the PR description.

**Comment by korydondzila** [comment]
@Exempt-Medic thanks for pointing that out, added more info.

File: worlds/shivers/Options.py
Comment (Exempt-Medic) [inline worlds/shivers/Options.py:34]: Should be able to remove this word. You could also change it to "getting access" or "obtaining access"
```suggestion
    For example: The Red Door puzzle will be logically required only after access to Beth's Address Book
```

File: worlds/shivers/Rules.py
Comment (Exempt-Medic) [inline worlds/shivers/Rules.py:79]: This should use `has_all`

File: worlds/shivers/Rules.py
Comment (Exempt-Medic) [inline worlds/shivers/Rules.py:89]: Do you really need to calculate `years_since_sep_30_1980` every time the completion condition is checked?

File: worlds/shivers/Rules.py
Comment (Exempt-Medic) [inline worlds/shivers/Rules.py:201]: These can all use the helper/convenience methods instead.
This applies to quite a few places in multiple files
```suggestion
        world.get_entrance(entrance_name).access_rule = rule
```

File: worlds/shivers/__init__.py
Comment (Exempt-Medic) [inline worlds/shivers/__init__.py:176]: This can just be `else` no?
```suggestion
            else:
```

File: worlds/shivers/__init__.py
Comment (Exempt-Medic) [inline worlds/shivers/__init__.py:166]: I'm confused why this only has a 50% chance of changing the value?

File: worlds/shivers/__init__.py
Comment (Exempt-Medic) [inline worlds/shivers/__init__.py:182]: Should be able to use `elif`
```suggestion
        elif self.options.lobby_access == 2:
```

File: worlds/shivers/__init__.py
Comment (Exempt-Medic) [inline worlds/shivers/__init__.py:195]:
```suggestion
        elif self.options.location_pot_pieces == "different_world":
```

File: worlds/shivers/__init__.py
Comment (Exempt-Medic) [inline worlds/shivers/__init__.py:141]: What you check for is inconsistent
```suggestion
            [loc for loc in library_region.locations if not loc.name.startswith("Storage: ")]
```

File: worlds/shivers/__init__.py
Comment (Exempt-Medic) [inline worlds/shivers/__init__.py:156]:
```suggestion
                 not loc.name.startswith("Storage: ") and loc != library_location]
```

File: worlds/shivers/__init__.py
Comment (Exempt-Medic) [inline worlds/shivers/__init__.py:249]: You should be doing this when you create the location/item. `pre_fill` can run into problems with plando

**Review by Exempt-Medic — CHANGES_REQUESTED** [review]
Some smaller comments, the changes overall look good to me though.

File: worlds/shivers/__init__.py
Comment (korydondzila) [reply worlds/shivers/__init__.py:166]: Ah in the case that early or local lobby access is used and front door is also usable, then we give a 50/50 chance that either the player gets the original 3 key route or the front door.

File: worlds/shivers/Rules.py
Comment (korydondzila) [reply worlds/shivers/Rules.py:201]: Ah didn't know about the convenience methods, that does clean up a fair bit.

File: worlds/shivers/Rules.py
Comment (korydondzila) [reply worlds/shivers/Rules.py]: Moved to constants

File: worlds/shivers/__init__.py
Comment (Exempt-Medic) [inline worlds/shivers/__init__.py:2]: You can combine these
```suggestion
from typing import Dict, List, Optional
```

File: worlds/shivers/__init__.py
Comment (Exempt-Medic) [inline worlds/shivers/__init__.py:180]: These should use strings
```suggestion
        if self.options.lobby_access == "early":
```

File: worlds/shivers/__init__.py
Comment (Exempt-Medic) [inline worlds/shivers/__init__.py:182]:
```suggestion
        elif self.options.lobby_access == "local":
```

Comment (Exempt-Medic) [inline worlds/shivers/__init__.py:182]: It might be good to update the documentation of the option here as well, since it says "Keys are placed locally" but this is placing them locally *and* early

**Review by Exempt-Medic — APPROVED** [review]
Overall, the changes LGTM. Some small comments above and some of the world-specific changes GodlFire would definitely need to look at, but it's waiting on their approval anyway. Did 100 solo generations, 100 10-player generations and 100 Shivers+DS3/Witness generations. No errors (or warnings) were encountered in these generations. Playthrough seems to still be fully intact, did not compare region graphs or anything like that before and after but presumably Kory Dondzila and GodlFire would know about those anyway. Did not test for client operations with these changes, but again, presumably they would/did.

File: worlds/shivers/__init__.py
Comment (korydondzila) [reply worlds/shivers/__init__.py:2]: Thought I ran import optimization on the Shivers directory

File: worlds/shivers/__init__.py
Comment (korydondzila) [reply worlds/shivers/__init__.py:180]: Ideally yeah, I have refactors like this that I wanted to do I just didn't want to overload this PR (I already did this with ItemTypes)

Comment (Exempt-Medic) [reply worlds/shivers/__init__.py:180]: I think doing this for two lines is fine

File: worlds/shivers/__init__.py
Comment (korydondzila) [reply worlds/shivers/__init__.py:182]: I was confused by what you meant, but I see, yes that is local_early_items which is different from just local_items. Wonder if I should change the option name then. (Ironically using local or local_early realistically wouldn't make much of a difference here except that the office elevator key and office keys could be placed very slightly later, I should discuss this with GodlFire)

File: worlds/shivers/__init__.py
Comment (korydondzila) [reply worlds/shivers/__init__.py:180]: Oh yeah the strings for sure, I was just responding on the "something that isn't just magic numbers", because also just using strings like this isn't ideal though it is better, enumeration would make more sense for options to me.

**Comment by Exempt-Medic** [comment]
This is out-of-scope, but I think it would make sense to eventually convert the item_table or at least create smaller tables for the item types so that something like:
```py
[self.create_item(name) for name, data in item_table.items() if data.type == ItemType.POT]
```
Could be:
```py
[self.create_item(name) for name in pot_item_table]
```

**Comment by korydondzila** [comment]
> This is out-of-scope, but I think it would make sense to eventually convert the item_table or at least create smaller tables for the item types so that something like:
> 
> ```python
> [self.create_item(name) for name, data in item_table.items() if data.type == ItemType.POT]
> ```
> 
> Could be:
> 
> ```python
> [self.create_item(name) for name in pot_item_table]
> ```

Yep

File: worlds/shivers/Rules.py
Comment (GodlFire) [inline worlds/shivers/Rules.py:174]: "Puzzle Solved Maze Door": lambda state: state.has("Viewed Theater Movie", player)

**Review by GodlFire — CHANGES_REQUESTED** [review]
World Maintainer:

As already stated Kory and I have already discussed all of these changes. Looking at the changes the region graphs do look correct Exempt-Medic, as does all of the rule changes.

I have also run a seed and it went fine.

I only see this one rule that could be changed to use the new events.

File: worlds/shivers/Rules.py
Comment (korydondzila) [reply worlds/shivers/Rules.py]: Done

**Review by GodlFire — APPROVED** [review]

File: worlds/shivers/Constants.py
Comment (Exempt-Medic) [inline worlds/shivers/Constants.py:19]: Is there a reason this uses relativedelta instead of something simple like:
```suggestion
years_since_sep_30_1980 = datetime.now().years - 1980
```

Comment (Exempt-Medic) [reply worlds/shivers/Constants.py:19]: Or maybe something like:
```py
datetime.now().year - 1980 + (datetime.now().month < 10)
```

Comment (korydondzila) [reply worlds/shivers/Constants.py:19]: Sep 30 1980 is a very specific date in the game. It's the date that the mystery officially occurs, and the game takes place in 1995, sometime in October or November (The game was released in November) though not specified.

image
image

So the victory item is basically an in joke where you've finally solved the mystery... X years later from that date. Relative delta just for it's accuracy honestly. (Though sadly the item name won't change if the game starts before Sep 30th but ends after Sep 30th).

File: worlds/shivers/Items.py
Comment (NewSoupVi) [inline worlds/shivers/Items.py:189]: Wait a minute, no.

Your location_name_to_id cannot change from one day to the next like this.
It is static "per run of AP", yes, but that is not enough. It has to be static across a *frozen build* of AP, no matter which day you run that build on.

The website loads location_name_to_id at different points of time / on different schedules for different purposes. Some things only happen once per **update** (If we didn't have the server crashes, options pages would still use the datapackage from **the release date of 0.5.0**), whereas for *generation*, a new instance of AP is loaded.

If you really want this, you'll have to make different items per year somehow. E.g. reserve like 100 of them, ahead of time, thinking either AP will be dead by then, or we'll have a more elegant solution to this datapackage issue.

Comment (NewSoupVi) [reply worlds/shivers/Items.py:189]: I have been informed this may be an event item. If it is code = None, then ignore everything I said.

Comment (NewSoupVi) [reply worlds/shivers/Items.py:189]: The item does appear to have a code in its ItemData, which makes me assume it's a real item

Comment (korydondzila) [reply worlds/shivers/Items.py:189]: Ah shit really? And here I was thinking I was doing something fun. What would that actually affect / break (In the window of being around Sep 30th each year)? I don't really know how everything works.

Hmm like I know it affects the web host in this way, though ugh I wish some items could be explicitly removed from these lists (since plenty of games have items that shouldn't be plando but aren't events).
Screenshot 2024-10-18 at 22 34 14

Because I could generate a bunch of these items with static years and the generation picks the one it wants to use at run time. (I mean could that really just be restricted to a couple years, since if it's too old it just falls off the list and too new then it won't matter. i.e. current +-1 year 43 44 and 45, then when next year rolls around it's 44 45 and 46?)

But again the webhost showing those is annoying.

And I know items show up in the tracker if received, but my assumption is that comes from whatever the generation said?

Comment (korydondzila) [reply worlds/shivers/Items.py:189]: @NewSoupVi didn't see the other messages, this is not an even item, it could be but the point was to have it show up when the player goals. So in any case it's never an item that will be randomized

Comment (NewSoupVi) [reply worlds/shivers/Items.py:189]: You can override options like LocalItems and provide your own set of `valid_keys`.

(In fact, if you removed this item from all `valid_keys` of all generic options, I could see an argument that *maybe* it's fine to have an item that changes names? I'm not sure though 🥴 I'd feel much more comfortable if you just reserved 100 different items or so ahead of time)

Comment (korydondzila) [reply worlds/shivers/Items.py:189]: @NewSoupVi Ah, I'll have to look into doing that, because we wouldn't want them in those options. What about the rolling +-1 item buffer since this is an item that only changes yearly we wouldn't really ever need more than 3 at a time.

Comment (NewSoupVi) [reply worlds/shivers/Items.py:189]: Hmm... a rolling item buffer *might* be okay. The reason I'm a bit wary is because it would still mean that some parts of the website would be working with:

```
43 years ago: ####0001
44 years ago: ####0002
45 years ago: ####0003
```

where other parts would be working with

```
44 years ago: ####0001
45 years ago: ####0002
46 years ago: ####0003
```

Because yes, this would mean that any name that would actually be used is always in there as long as there is a server restart at least once a year (lol), but which name belongs to which ID would be mismatched.
I don't *think* that causes any issues, but I'm not sure. I'd have to verify with other core devs

Comment (korydondzila) [reply worlds/shivers/Items.py:189]: I can associate the id with the year 

```
43 years ago: ####00043
44 years ago: ####00044
45 years ago: ####00045
```

```
44 years ago: ####00044
45 years ago: ####00045
46 years ago: ####00046
```

Comment (NewSoupVi) [reply worlds/shivers/Items.py:189]: Oh sure. At that point I don't really see the benefit over just reserving a bunch ahead of time, other than I guess "it works forever", which I don't imagine will be relevant but it is *technically* slightly safer. Lol

Comment (korydondzila) [reply worlds/shivers/Items.py:189]: I'm both lazy and don't want a bunch of extra stuff laying around that won't get used lol.

Comment (korydondzila) [reply worlds/shivers/Items.py]: @NewSoupVi ok not that I had time to do this, done. Have my rolling 3 item buffer and removed them from valid keys so they don't show on web.

File: worlds/shivers/Options.py
Comment (Exempt-Medic) [inline worlds/shivers/Options.py:7]: These need to be relative imports
```suggestion
from . import ItemType, item_table
from .Constants import location_info
```

File: worlds/shivers/Options.py
Comment (Exempt-Medic) [inline worlds/shivers/Options.py:127]: This whole process doesn't really fix everything. Plando still exists, among other things, so it's still going to run into problems, it's also kind of "hacky" to overwrite all of these. Can you really not do anything else?

Comment (korydondzila) [reply worlds/shivers/Options.py:127]: I mean this is what was suggested, if there's a better way I would love to know.

Comment (Exempt-Medic) [reply worlds/shivers/Options.py:127]: Just using an event?

Comment (Exempt-Medic) [reply worlds/shivers/Options.py:127]: As far as I can tell, the only thing this is being used for is the completion_condition, so I'm not sure why this isn't just an event (ID None)

Comment (korydondzila) [reply worlds/shivers/Options.py:127]: I think I mentioned it before but I was trying to be weird and have an in joke goal item that displays when you beat the game. Event items will not do that.
 
However something like the Map Complete items in the Dooms/Heretic would be similar and display in these options when they probably should not. afaik this is the only game so far attempting do this, but nothing says it can't. Plando is of course still a problem, but that requires deeper knowledge, the web based options is at least something that can be mitagated.

Comment (Exempt-Medic) [reply worlds/shivers/Options.py:127]: Can't you just make the client display the item anyway? Just fake one, basically?

Comment (Exempt-Medic) [reply worlds/shivers/Options.py:127]: Idk, I'll still leave my approval, but I'm starting to like less and less all the weird things worlds are doing with non-event "events" XD

Comment (korydondzila) [reply worlds/shivers/Options.py:127]: At least this cleans up all the other stuff that should actually have been events but weren't and added events to clean up rule logic. I just want my one small weird thing lol.

File: worlds/shivers/requirements.txt
Comment (Berserker66) [inline worlds/shivers/requirements.txt:1]: adding this requirement for calculating the years since a date seems... excessive?

Comment (korydondzila) [reply worlds/shivers/requirements.txt]: removed

File: worlds/shivers/Constants.py
Comment (korydondzila) [reply worlds/shivers/Constants.py]: Removed dateutil

**Review by NewSoupVi — APPROVED** [review]

**Comment by NewSoupVi** [comment]
I did not mean to do that last thing lol

**Comment by NewSoupVi** [comment]
Anyway, after merging this, Exempt-Medic reminded me that there was an outstanding issue with this PR that was not explicitly noted down as a requested change:

Your LocalItems class should probably subclass Options.LocalItems

It's not that bad but I would appreciate a followup PR changing that

**Comment by korydondzila** [comment]
> Anyway, after merging this, Exempt-Medic reminded me that there was an outstanding issue with this PR that was not explicitly noted down as a requested change:
> 
> Your LocalItems class should probably subclass Options.LocalItems instead of Options.ItemSet
> 
> It's not that bad but I would appreciate a followup PR changing that

Yeah I was going to work on that. at least that will be a small simple PR.

**Pull request merged**
